_posts/2021-02-11-please-welcome-attribute-routing-to-servicecomposer.md

```diff
@@ -0,0 +1,142 @@
+
+> What's ServiceComposer? ServiceComposer is a view model composition gateway designed to compose data owned by different (micro)services and transparently serve users' requests with a single response view model. More information about the overall architectural problem and the many nuances are available in the [ViewModel Composition series](https://milestone.topics.it/categories/view-model-composition) of articles on this blog.
+
+With the ability to use routing standalone, we can now write something like:
+
+```csharp
+public class Startup
+{
+    public void ConfigureServices(IServiceCollection services)
+    {
+        services.AddRouting();
+    }
+
+    public void Configure(IApplicationBuilder app, ILoggerFactory loggerFactory)
+    {
+        var routeBuilder = new RouteBuilder(app);
+        routeBuilder.MapRoute("{controller}/{id:int?}", context =>
+        {
+            //handle the incoming HTTP Request here
+            return Task.CompletedTask;
+        });
+        app.UseRouter(routeBuilder.Build());
+    }
+}
+```
+
+As you can see, it's pretty low-level stuff; you can barely use route templates and map those to callback functions. It was what ServiceComposer was doing under the hood. On top of that, in ServiceComposer, we built a higher-level API to abstract away some of the complexity.
+
+## Problem: Lack of authentication and authorization support
+
+Requests to support authentication and authorization came with no surprise (for example [here](https://github.com/ServiceComposer/ServiceComposer.AspNetCore.Mvc/issues/53) and [here](https://github.com/mauroservienti/designing-a-ui-for-microservices-demos/issues/2)). The problem was that to provide an authentication and authorization model; we'd have had to write it from scratch.
+
+To handle a request and compose results using ServiceComposer, users need to define request composition handlers, something like the following:
+
+```csharp
+class SampleHandler : ICompostionRequestHandler
+{
+   public Task Handle(HttpRequest request)
+   {
+      //composition steps here.
+   }
+}
+```
+
+> A request composition handler is conceptually similar to an MVC Controller. The main difference is that in MVC, controllers are a one to one association with routes. One controller can serve only one route, and one route can have only one controller. In ServiceComposer, we can associate many request handlers with one route to allow a cooperative data composition.
+
+To elegantly solve the authorization and authentication issue, it would have been great to be able to allow the usage of regular ASP.Net authentication and authorization attributes, e.g.:
```

Comment:
```suggestion
To elegantly solve the authorization and authentication issue, it would have been great to allow the use of regular ASP.NET authentication and authorization attributes, e.g.:
```